fendermint/vm/interpreter/src/fvm/examples/mycustomkernel.rs

```diff
@@ -0,0 +1,201 @@
+use fvm_shared::consensus::ConsensusFault;
+use fvm_shared::piece::PieceInfo;
+use fvm_shared::randomness::RANDOMNESS_LENGTH;
+use fvm_shared::sector::{
+    AggregateSealVerifyProofAndInfos, RegisteredSealProof, ReplicaUpdateInfo, SealVerifyInfo,
+};
+use fvm_shared::sys::out::network::NetworkContext;
+use fvm_shared::sys::out::vm::MessageContext;
+use fvm_shared::{address::Address, econ::TokenAmount, ActorID, MethodNum};
+
+use ambassador::Delegate;
+use cid::Cid;
+
+// we define a single custom syscall which simply doubles the input
+pub trait CustomKernel: Kernel {
+    fn my_custom_syscall(&self, doubleme: i32) -> Result<i32>;
+}
+
+// our custom kernel extends the filecoin kernel
+#[derive(Delegate)]
+#[delegate(IpldBlockOps, where = "C: CallManager")]
+#[delegate(ActorOps, where = "C: CallManager")]
+#[delegate(CryptoOps, where = "C: CallManager")]
+#[delegate(DebugOps, where = "C: CallManager")]
+#[delegate(EventOps, where = "C: CallManager")]
+#[delegate(MessageOps, where = "C: CallManager")]
+#[delegate(NetworkOps, where = "C: CallManager")]
+#[delegate(RandomnessOps, where = "C: CallManager")]
+#[delegate(SelfOps, where = "C: CallManager")]
+#[delegate(SendOps<K>, generics = "K", where = "K: CustomKernel")]
+#[delegate(UpgradeOps<K>, generics = "K", where = "K: CustomKernel")]
+pub struct DefaultCustomKernel<C>(pub DefaultFilecoinKernel<C>);
+
+impl<C> CustomKernel for DefaultCustomKernel<C>
+where
+    C: CallManager,
+    DefaultCustomKernel<C>: Kernel,
+{
+    fn my_custom_syscall(&self, doubleme: i32) -> Result<i32> {
+        Ok(doubleme * 2)
+    }
+}
+
+impl<C> DefaultCustomKernel<C>
+where
+    C: CallManager,
+{
+    fn price_list(&self) -> &PriceList {
+        (self.0).0.call_manager.price_list()
+    }
+}
+
+impl<C> Kernel for DefaultCustomKernel<C>
+where
+    C: CallManager,
+{
+    type CallManager = C;
+    type Limiter = <DefaultFilecoinKernel<C> as Kernel>::Limiter;
+
+    fn into_inner(self) -> (Self::CallManager, BlockRegistry)
+    where
+        Self: Sized,
+    {
+        self.0.into_inner()
+    }
+
+    fn new(
+        mgr: C,
+        blocks: BlockRegistry,
+        caller: ActorID,
+        actor_id: ActorID,
+        method: MethodNum,
+        value_received: TokenAmount,
+        read_only: bool,
+    ) -> Self {
+        DefaultCustomKernel(DefaultFilecoinKernel::new(
+            mgr,
+            blocks,
+            caller,
+            actor_id,
+            method,
+            value_received,
+            read_only,
+        ))
+    }
+
+    fn machine(&self) -> &<Self::CallManager as CallManager>::Machine {
+        self.0.machine()
+    }
+
+    fn limiter_mut(&mut self) -> &mut Self::Limiter {
+        self.0.limiter_mut()
+    }
+
+    fn gas_available(&self) -> Gas {
+        self.0.gas_available()
+    }
+
+    fn charge_gas(&self, name: &str, compute: Gas) -> Result<GasTimer> {
+        self.0.charge_gas(name, compute)
+    }
+}
+
+impl<C> FilecoinKernel for DefaultCustomKernel<C>
```

Comment:
@fridrik01 since are customizing syscall for IPC, is it necessary to implement FilcoinKernel for DefaultCustomKernel?

Reply:
nope, will update!